File: 2022/day_21.cpp
//
// Created by Wollenschneider Luis on 03.12.22.
//

#include "../utils.hpp"

#define DAY "21"
#define YEAR "2022"
#define RETURN_TYPE long long int

struct Operation {
    char op;
    std::string arg1;
    std::string arg2;
    RETURN_TYPE res;
};

#define TYPE std::map<std::string, Operation>

TYPE parse_input(const std::string &filepath) {
    // check if path exists
    TYPE input;
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cout << "File not found" << std::endl;
        return input;
    }
    std::string line;
    while (std::getline(file, line)) {
        Operation op;
        std::string name = line.substr(0, line.find(':'));
        line = line.substr(line.find(':') + 2);
        if (line.length() == 11) {
            op.op = line[5];
            op.arg1 = line.substr(0, 4);
            op.arg2 = line.substr(7, 4);
            op.res = -1;
        } else {
            op.op = 0;
            op.arg1 = "";
            op.arg2 = "";
            op.res = std::stoi(line);
        }
        input[name] = op;
    }
    file.close();
    return input;
}

RETURN_TYPE calculate(TYPE &inp, const std::string& ins) {
    if (inp[ins].res != -1) {
        return inp[ins].res;
    }
    RETURN_TYPE a = calculate(inp, inp[ins].arg1);
    RETURN_TYPE b = calculate(inp, inp[ins].arg2);
    RETURN_TYPE res = 0;
    switch (inp[ins].op) {
        case '+': {
            res = a + b;
            break;
        }
        case '*': {
            res = a * b;
            break;
        }
        case '-': {
            res = a - b;
            break;
        }
        case '/': {
            res = a / b;
            break;
        }
        case '=' : {
            res = a == b;
            break;
        }
    }
    return res;
}


RETURN_TYPE part1(TYPE &inp) {
    return calculate(inp, "root");
}

RETURN_TYPE pre_compute(TYPE &inp, const std::string& ins) {
    if (ins == "humn") {
        return -1;
    }
    if (inp[ins].res != -1) {
        return inp[ins].res;
    }
    RETURN_TYPE a = pre_compute(inp, inp[ins].arg1);
    RETURN_TYPE b = pre_compute(inp, inp[ins].arg2);
    if (a == -1 || b == -1) {
        return -1;
    }
    RETURN_TYPE res = 0;
    switch (inp[ins].op) {
        case '+': {
            res = a + b;
            break;
        }
        case '*': {
            res = a * b;
            break;
        }
        case '-': {
            res = a - b;
            break;
        }
        case '/': {
            res = a / b;
            break;
        }
        case '=' : {
            res = a == b;
            break;
        }
    }
    inp[ins].res = res;
    return res;
}

RETURN_TYPE backtrace(TYPE &inp, const std::string& ins, RETURN_TYPE res) {
    if (ins == "humn") {
        return res;
    }
    RETURN_TYPE a = inp[inp[ins].arg1].res;
    RETURN_TYPE b = inp[inp[ins].arg2].res;
    std::string instruction;
    RETURN_TYPE res2;
    if (a == -1) {
        instruction = inp[ins].arg1;
        switch (inp[ins].op) {
            case '+': {
                // res = a + b;
                res2 = res - b;
                break;
            }
            case '*': {
                // res = a * b;
                res2 = res / b;
                break;
            }
            case '-': {
                // res = a - b;
                res2 = res + b;
                break;
            }
            case '/': {
                // res = a / b;
                res2 = res * b;
                break;
            }
            case '=' : {
                res2 = b;
                break;
            }
        }
    } else if (b == -1) {
        instruction = inp[ins].arg2;
        switch (inp[ins].op) {
            case '+': {
                // res = a + b;
                res2 = res - a;
                break;
            }
            case '*': {
                // res = a * b;
                res2 = res / a;
                break;
            }
            case '-': {
                // res = a - b;
                res2 = a - res;
                break;
            }
            case '/': {
                // res = a / b;
                res2 = a / res;
                break;
            }
            case '=' : {
                res2 = a;
                break;
            }
        }
    } else {
        std::cout << "Error" << std::endl;
        exit(1);
    }
    return backtrace(inp, instruction, res2);
}

RETURN_TYPE part2(TYPE &inp) {
    inp["root"].op = '=';
    pre_compute(inp, "root");
    inp["humn"].res = -1;

    return backtrace(inp, "root", 1);
}

int main() {
    TYPE test_inp = parse_input(YEAR "/tests/day_" DAY ".txt");
    TYPE inp = parse_input(YEAR "/inputs/day_" DAY ".txt");

    auto expected_result_part1 = get_expected_result<RETURN_TYPE>(YEAR "/tests/results/day_" DAY "_1.txt");

    RETURN_TYPE res = part1(test_inp);
    bool eval_res = evaluate_results(res, expected_result_part1);
    res = part1(inp);
    std::set<RETURN_TYPE> tried;
    tried.insert(-393342880);
    tried_before(res, tried);
    std::cout << "Part 1: " << res << std::endl;

    if (!eval_res) {
        return 1;
    }

    std::cout << std::endl;
    auto expected_result_part2 = get_expected_result<RETURN_TYPE>(YEAR "/tests/results/day_" DAY "_2.txt");

    res = part2(test_inp);
    evaluate_results(res, expected_result_part2);
    res = part2(inp);
    std::set<RETURN_TYPE> tried2;
    tried2.insert(10708574);
    tried_before(res, tried2);
    std::cout << COLOR_RED "Written in Python, due to overflow!" COLOR_RESET << std::endl;
    std::cout << "Part 2: " << res << std::endl;

    return 0;
}

File: 2022/day_21.py
from utils import aoc_comm, create_test_file
import os
from typing import Optional
import re

# --- update day/ year for each challenge
settings = {
    'day': 21,
    'year': 2022,
    'cookie-path': os.path.realpath('../aoc_cookie.json')
}

# -- OTHER LIBs that might help while coding the solutions
from collections import Counter, defaultdict
from math import *
import functools
import itertools

class Op:
    arg1 = None
    arg2 = None
    op = None
    res = None


def parse_input(inp_content):
    inp_content = inp_content.strip()
    # add further input processing here..
    inp_content = [e for e in inp_content.split("\n")]
    inp = {}
    for line in inp_content:
        o = Op()
        try:
            o.res = int(line.split(": ")[1].strip())
        except ValueError:
            o.op = line[11]
            o.arg1 = line[6:10]
            o.arg2 = line[13:17]
        inp[line.split(": ")[0]] = o
    return inp

    
@aoc_comm(settings, level=1)
def solve_l1(input_str) -> Optional[int]:  # input data will be passed to this as string
    inp = parse_input(input_str)

    ans = None
    for ee in inp:
        pass

    return ans  # if 'ans' is None answer won't be submitted, else it will submit after confirmation


def pre_compute(inp, ins):
    if ins == "humn":
        return None
    if inp[ins].res is not None:
        return inp[ins].res
    a = pre_compute(inp, inp[ins].arg1)
    b = pre_compute(inp, inp[ins].arg2)
    if a is None or b is None:
        return None
    res = 0
    if inp[ins].op == "+":
        res = a + b
    elif inp[ins].op == "*":
        res = a * b
    elif inp[ins].op == "-":
        res = a - b
    elif inp[ins].op == "/":
        res = a // b
    elif inp[ins].op == "=":
        res = 1 if a == b else 0
    inp[ins].res = res
    return res


def backtrace(inp, ins, res):
    if ins == "humn":
        return res
    a = inp[inp[ins].arg1].res
    b = inp[inp[ins].arg2].res
    if a is None:
        if inp[ins].op == "+":
            a = res - b
        elif inp[ins].op == "*":
            a = res // b
        elif inp[ins].op == "-":
            a = res + b
        elif inp[ins].op == "/":
            a = res * b
        elif inp[ins].op == "=":
            a = b if res == 1 else 0
        return backtrace(inp, inp[ins].arg1, a)
    elif b is None:
        if inp[ins].op == "+":
            b = res - a
        elif inp[ins].op == "*":
            b = res // a
        elif inp[ins].op == "-":
            b = a - res
        elif inp[ins].op == "/":
            b = a // res
        elif inp[ins].op == "=":
            b = a if res == 1 else 0
        return backtrace(inp, inp[ins].arg2, b)
    else:
        print("ERROR")
        exit(1)


@aoc_comm(settings, level=2)
def solve_l2(input_str) -> Optional[int]:
    inp = parse_input(input_str)

    inp["root"].op = '='
    pre_compute(inp, "root")
    inp["humn"].res = None

    return backtrace(inp, "root", 1)


def main():
    # create_test_file(settings, int(input("Test file for level: ")), input("Test result: ").strip())
    solve_l1()
    solve_l2()


if __name__ == '__main__':
    main()
